### ISSUE TITLE
Introduce a groundtruths argument in the call method of tf.keras.layer

### ISSUE THREAD

**Issue by EmGarr** [tensorflow/tensorflow#39810]
**System information**: Ubuntu 16.04
- TensorFlow version (you are using): 2.2.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Today, the API only support three arguments (inputs, training, masks):

```python
class MyNetwork(tf.keras.layers.Layer):

  def call(self, inputs, training=None):  
     # Do somestuff
      return stuff
```

I propose to introduce a fourth one (groundtruths or other):

```python
class MyNetwork(tf.keras.layers.Layer):

  def call(self, inputs, training=None, groundtruths=None):  
     # Do somestuff
      return stuff
```

You could tell: why don't you put your groundtruths inside the inputs? It leads to issue with the SavedModel format. My inputs in training and inference won't be the same which leads to errors.

Example:

```python
training_inputs = {'image': tf.zeros((1, 800, 800, 3)), 'gt_bbox':  tf.constant([[0, 0, 1, 1]], tf.float32)}
inference_inputs = {'image': tf.zeros((1, 800, 800, 3))}
```
**Will this change the current api? How?**

It will change the interface of the `call` method from `tf.keras.layers.Layer`.

**Who will benefit with this feature?**

People working on algorithms like FasterRCNN need to pass during the training the groundtruths 
to perform sampling on it. Currently you need to perform lot of hacks to be able to export your model.

**Any Other info.**

Maybe adding new arguments to the call method isn't the right way. Allowing saved_model to understand that you have two types of signatures, when training is True or False, for your model would be great.  It would make the following behavior compatible to saved_model.

Example:

```python
class MyNetwork(tf.keras.Model):

  def call(self, inputs, training=None):  
     if training:
          groundtruths = inputs['groundtruths']
          # Do somestuff with the groundtruths
      return stuff

    def train_step(self, data):
        data = data_adapter.expand_1d(data)
        x, y, _ = data_adapter.unpack_x_y_sample_weight(data)

        with tf.GradientTape() as tape:
            x['ground_truths'] = y
            y_pred = self(x, training=True)
            loss = self.compiled_loss(None, y_pred, None, regularization_losses=self.losses)
        training._minimize(self.distribute_strategy, tape, self.optimizer, loss,
                           self.trainable_variables)
        return {m.name: m.result() for m in self.metrics}

```

**Comment by tanzhenyu** [comment]
@EmGarr Thanks for filing the request! This is a rather important feature which I think we should support. However special-casing an argument for detection models would make things too complicated. In general I don't think we can make that into keras.Layer. But likely into keras.layers.ImageLayer which inherits from it and deals with both image and gt boxes

**Comment by EmGarr** [comment]
@tanzhenyu I think the issue can be closed. It seems that the futur handling of nested arguments for a tf.function will solve this issue:
https://www.tensorflow.org/guide/concrete_function#changes_for_tensorflow_23.
